src/readpixels.web.js

```diff
@@ -19,6 +20,9 @@ function readpixels(url, P, limit = 0) {
   const ctx = canvas.getContext('2d')
 
   return new P((resolve, reject) => {
+    if (typeof url === 'object' && url.nodeName.toLowerCase() === 'canvas') {
+      return resolve(url.getContext('2d').getImageData(0, 0, url.width, url.height));
+    }
```

Comment:
We may want to add a method like in [readpixels.js](https://github.com/obartra/ssim/blob/master/src/readpixels.js#L112) (the node equivalent of `readpixels.web.js`), where based on the input type we take an action or another.

As parameter I would pass the output of `getImageData` directly though, that way it's not up to `ssim` to figure out the coordinates to retrieve.

Reply:
On your first point, have I understood you correctly that you're suggesting to refactor `readpixels()` to grab promises from other methods instead?

I don't understand your second point, could you please clarify?

Reply:
I was thinking something along the lines:

```js
/**
 * Reads image data from the input and returns it
 *
 * @method loadImage
 * @param {string} url - The url to use to load the image data
 * @param {function} P - The Promise definition, must be a valid Promises/A+ implementation
 * @param {number} [limit=0] - A limit that, if set and both dimensions (width / height) surpass it,
 * will downsize the image to that size on the smallest dimension.
 * @returns {Promise} promise - A promise that resolves with the image 3D matrix
 * @public
 * @memberOf readpixelsWeb
 */
function loadImage(url, P, limit = 0) {
  const img = new Image()
  const canvas = document.createElement('canvas')
  const ctx = canvas.getContext('2d')

  return new P((resolve, reject) => {
    img.onload = () => {
      const { width, height } = getLimitDimensions(img.width, img.height, limit)

      if (width === 0 || height === 0) {
        return reject('Failed to load image')
      }

      canvas.width = width
      canvas.height = height

      ctx.drawImage(img, 0, 0, img.width, img.height, 0, 0, width, height)
      return resolve(ctx.getImageData(0, 0, width, height))
    }
    img.onerror = reject
    img.src = url
  })
}

/**
 * Reads image data from the input and returns it
 *
 * @method readpixels
 * @param {string|ImageData} url - A url to load the image data or the imageData instance directly
 * @param {function} P - The Promise definition, must be a valid Promises/A+ implementation
 * @param {number} [limit=0] - A limit that, if set and both dimensions (width / height) surpass it,
 * will downsize the image to that size on the smallest dimension (unused when ImageData is passed directly)
 * @returns {Promise} promise - A promise that resolves with the image 3D matrix
 * @public
 * @memberOf readpixelsWeb
 * @since 0.0.1
 */
function readpixels(url, P, limit = 0) {
  if (url instanceof ImageData) {
    return P.resolve(url)
  }
  return loadImage(url, P, limit)
}
```

What that buys is that there's no need to handle the canvas from within ssim. With it you would call it like:

```js
function getImageData(canvas){
  return canvas.getContext('2d').getImageData(0, 0, url.width, url.height));
}

const originalImage = document.querySelector('#canvas-original')
const modifiedImage = document.querySelector('#canvas-modified')

ssim(getImageData(originalImage), getImageData(modifiedImage))
```

Reply:
Thanks for the clarification, however is that example code not different in structure to how the node.js specific readpixels implementation functions where the `readpixels()` function can accept certain types, and then farms the call out to another method which performs the actual loading? I would be concerened about the API between the two deviating too far.

I'm also unclear as to what your concern is with handling the canvas inside of ssim, unless you're proposing a much wider change impacting change - which I just wanted to get something relatively clean that suited my use case without breaking it for other people.

Reply:
This shouldn't be a big change! (just refactoring your code from one function to two, matching what `readpixels.js` does).

The concern with handling canvas inside ssim is that it makes assumptions about your canvas image (e.g. what if the imagedata you want to compare is about different areas in the same canvas element?) we can avoid that problem if we just pass the imagedata constructor.

In fact, looking at this codebase after 2 years I think it's trying to do too much, both in node and web, it would probably be best if it always just got the image data. In that sense if we end up making a new release I imagine the project moving in that direction so I wouldn't want it to also be responsible for web canvases